libcontainer/cgroups/fs/memory.go

```diff
@@ -96,9 +96,18 @@ func setKernelMemory(path string, kernelMemoryLimit int64) error {
 }
 
 func setMemoryAndSwap(path string, cgroup *configs.Cgroup) error {
-	// When memory and swap memory are both set, we need to handle the cases
-	// for updating container.
-	if cgroup.Resources.Memory != 0 && cgroup.Resources.MemorySwap > 0 {
+	// Memeory set to '-1' which is unlimited. In this situation we need to update swap to -1
+	// to avoid kernel validation failures.
+	if cgroup.Resources.Memory == -1 {
+		if err := writeFile(path, "memory.memsw.limit_in_bytes", strconv.FormatInt(-1, 10)); err != nil {
```

Comment:
libcontainer as a low level library usually don't do this kind of thing, it only does what it's told to do.

So if docker want to set memory as -1, it should pass memorySwap as -1 as well, libcontainer only adapt the writing sequence to make the logic right.